### PR TITLE
Add 2 new Markdown IT Functionality (include and container)

### PR DESCRIPTION
**Pull request by L3houx**
- Add documentation in README (about the 2 functionalities that I added)
- Update package.json
- Modify src/System/Tasks/ConversionRunner.ts to add 2 Markdown-It package (include and container)
- Add a declare file to be able to compile the project du to the fact that markdown-it-include didn't had a @type

Thank You and Have a nice one!

### CODE REVIEW

File: src/System/Tasks/ConversionRunner.ts [hunk 1]
```diff
@@ -6,6 +6,10 @@ import cloneDeep = require("lodash.clonedeep");
 import MarkdownIt = require("markdown-it");
 import checkbox = require("markdown-it-checkbox");
 import emoji = require("markdown-it-emoji");
+import markdownContainer = require("markdown-it-container");
+import markdownInclude = require("markdown-it-include");
+import path = require("path");
```
Comment (manuth) [inline src/System/Tasks/ConversionRunner.ts:11]: Please remove this and use `upath`s `dirname` instead. It's already imported on line 15:
https://github.com/manuth/MarkdownConverter/blob/f3dde71d9455746939e230887f9ef05b60c9c1f0/src/System/Tasks/ConversionRunner.ts#L15

File: src/System/Tasks/ConversionRunner.ts [hunk 2]
```diff
@@ -489,6 +493,36 @@ export class ConversionRunner
...
+        /**
+         * Add Markdown IT Include
+         */
+
+        var markdownIncludeOptions = {
+            root: path.dirname(window.activeTextEditor.document.fileName),
+            includeRe: /\!{3}\s*include(.+?)\!{3}/i
+          };
+
+        parser.use(markdownInclude, markdownIncludeOptions);
```
Comment (manuth) [inline src/System/Tasks/ConversionRunner.ts:523]: Please do not use the `var` keyword. Instead just pass the options directly:
```suggestion
        parser.use(
            markdownContainer,
            '',
            {
                validate: function (name: string) {
                    return name.trim().length;
                },
                render: function (token:Token[], id:number) {
                    if (token[id].info.trim() !== '') {
                        return `<div class="${token[id].info.trim()}">\n`;
                    } else {
                        return `</div>\n`;
                    }
                }
          });

        /**
         * Add Markdown IT Include
         */
        parser.use(
            markdownInclude,
            {
                root: path.dirname(window.activeTextEditor.document.fileName),
                includeRe: /\!{3}\s*include(.+?)\!{3}/i
            });
```

Comment (manuth) [inline src/System/Tasks/ConversionRunner.ts:523]: Please do not use the `var` keyword. Instead just pass the options directly:
```suggestion
        parser.use(
            markdownContainer,
            '',
            {
                validate: function (name: string) {
                    return name.trim().length;
                },
                render: function (token:Token[], id:number) {
                    if (token[id].info.trim() !== '') {
                        return `<div class="${token[id].info.trim()}">\n`;
                    } else {
                        return `</div>\n`;
                    }
                }
          });

        /**
         * Add Markdown IT Include
         */
        parser.use(
            markdownInclude,
            {
                root: path.dirname(window.activeTextEditor.document.fileName),
                includeRe: /\!{3}\s*include(.+?)\!{3}/i
            });
```

File: src/System/Tasks/ConversionRunner.ts [hunk 3]
```diff
@@ -489,6 +493,36 @@ export class ConversionRunner
...
+
+
```
Comment (manuth) [inline src/System/Tasks/ConversionRunner.ts:525]: Please remove the duplicate new line character

File: package-lock.json [hunk 4]
```diff
@@ -1,214 +1,99 @@
 {
   "name": "markdown-converter",
   "version": "5.0.1",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Comment (manuth) [inline package-lock.json:4]: Please install at least npm v7, completely revert all changes made to this file (`git checkout package-lock.json origin/dev`) and re-run `npm install`.
Having this `package-lock.json` file generated using `npm` with a version < 7 breaks the CI process (as npm v6 and lower **doesn't** install the `peerDependencies` of `dependencies` automatically while npm v7 and higher does).

I assume this is the reason for the CI process to fail.

**Review by manuth — CHANGES_REQUESTED** [review]
That's awesome! I've been looking for a convenient way to include other .md files and to apply CSS classes for long, tbh.

Before we can get started w/ merging this PR, please take note of the comments I've made and change your code appropriately.

In order to have your code checked correctly, you'll have to fix your code.

I assume the error that the CI is facing is caused by the `package-lock.json` file being generated using an `npm` version which is lower than v7.

Please perform following steps in order to fix it:
  - Revert changes made to `package-lock.json`: `git checkout origin/dev -- package-lock.json`
  - Install npm v7 (or higher - you might have to install a more recent version of nodejs in order to do so)
  - Re-run `npm install`

For a next time you might want to keep seperate features truly separate (i.e. create a separate PR for the CSS feature and a separate PR for the Include feature).

It's okay, though - we'll leave it as is for now.

Thanks a lot for the effort you put into this, I really appreciate it ✨

File: README.md [hunk 5]
```diff
@@ -8,18 +8,18 @@ A markdown-converter for [vscode][vscode]
...
     - [Date-Formatting](#date-formatting)
-      - [Custom Date-Formats](#custom-date-formats)
   - [Headers and Footers](#headers-and-footers)
   - [Including Table of Contents](#including-table-of-contents)
+  - [Include Markdown Fragment Files](#include-markdown-fragment-files)
+  - [Creating Block-Level Custom Containers](#creating-block-level-custom-containers)
   - [Third Party Markdown Extensions](#third-party-markdown-extensions)
   - [Assets](#assets)
-    - [CSS- and JavaScript-Files](#css--and-javascript-files)
-      - [Insertion](#insertion)
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 6]
```diff
@@ -8,18 +8,18 @@ A markdown-converter for [vscode][vscode]
...
+    - [CSS- and JavaScript-Files](#css-and-javascript-files)
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 7]
```diff
@@ -8,18 +8,18 @@ A markdown-converter for [vscode][vscode]
   - [Usage](#usage)
   - [Variable Substitution](#variable-substitution)
     - [Date-Formatting](#date-formatting)
-      - [Custom Date-Formats](#custom-date-formats)
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 8]
```diff
@@ -8,18 +8,18 @@ A markdown-converter for [vscode][vscode]
...
   - [Assets](#assets)
-    - [CSS- and JavaScript-Files](#css--and-javascript-files)
-      - [Insertion](#insertion)
+    - [CSS- and JavaScript-Files](#css-and-javascript-files)
     - [Pictures](#pictures)
   - [Settings](#settings)
 
 ## What is `MarkdownConverter`?
-> MarkdownConverter is a Visual Studio Code-extension which allows you to export your Markdown-file as PDF-, HTML or Image-files.  
+> MarkdownConverter is a Visual Studio Code-extension which allows you to export your Markdown-file as PDF-, HTML or Image-files.
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 9]
```diff
@@ -42,14 +42,14 @@ A markdown-converter for [vscode][vscode]
  3. Open up the command pallet (<kbd>Ctrl</kbd>, <kbd>Shift</kbd>+<kbd>P</kbd>) and search one of these commands:
     - `Markdown: Convert Document` (`Markdown: Dokument Konvertieren` in German) or `mco` (`mk` in German) for short
     - `Markdown: Convert all Documents` (`Markdown: Alle Dokumente konvertieren`) or `mcd` (`madk` in German) for short
-    - `Markdown: Chain all Documents` (`Markdown: Alle Dokumente verketten`) or `mcad` (`madv` in German) for short 
+    - `Markdown: Chain all Documents` (`Markdown: Alle Dokumente verketten`) or `mcad` (`madv` in German) for short
  4. Press enter and wait for the process to finish
 
-Normally, `MarkdownConverter` will refuse to convert files which aren't recognized as markdown-files.  
+Normally, `MarkdownConverter` will refuse to convert files which aren't recognized as markdown-files.
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 10]
```diff
@@ -42,14 +42,14 @@ A markdown-converter for [vscode][vscode]
...
 If you don't like this behavior, you might want to set `markdownConverter.IgnoreLanguageMode` to `true`.
 
 ## Variable Substitution
-Before the conversion, the markdown-file is preprocessed using [`Handlebars`][Handlebars]. Variables (such as `{{ Author }}`) are automatically replaced with the corresponding attribute-value.  
+Before the conversion, the markdown-file is preprocessed using [`Handlebars`][Handlebars]. Variables (such as `{{ Author }}`) are automatically replaced with the corresponding attribute-value.
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 11]
```diff
@@ -67,15 +67,15 @@ This page was written by {{ Author }}
...
+  * `CreationDate`
     A `Date`-value indicating the time of the creation of the markdown-file
-  * `ChangeDate`  
+  * `ChangeDate`
     A `Date`-value indicating the time of the last change of the content of the markdown-file
-  * `CurrentDate`  
+  * `CurrentDate`
     A `Date`-value representing the time of the conversion
-  * `Author`  
+  * `Author`
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 12]
```diff
@@ -172,7 +172,7 @@ This is a test.
 If you'd like to have more control on what your headers and footers look like, you might want to adjust the `markdownConverter.Document.HeaderTemplate` and `markdownConverter.Document.FooterTemplate` options.
 
 ## Including Table of Contents
-The `MarkdownConverter` provides the functionality to create a table of contents at runtime.  
+The `MarkdownConverter` provides the functionality to create a table of contents at runtime.
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 13]
```diff
@@ -191,6 +191,29 @@ Furthermore, the `markdownConverter.Parser.Toc.Levels` allows you to choose whic
 }
 ```
 
+## Include Markdown Fragment Files
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 14]
```diff
@@ -191,6 +191,29 @@ Furthermore, the `markdownConverter.Parser.Toc.Levels` allows you to choose whic
...
+
+:::
+```
+
+In CSS File:
+```
+div.class {
+   Custom CSS Here
+ }
+```
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: README.md [hunk 15]
```diff
@@ -259,38 +282,38 @@ The `markdownConverter.Assets.PictureInsertion`-option allows you to set whether
...
+  - `markdownConverter.Document.HeaderContent` and `markdownConverter.Document.FooterContent`:
     Allows you to set the content of the different sections of the header and the footer.
-  - `markdownConverter.Document.HeaderTemplate` and `markdownConverter.Document.FooterTemplate`:  
+  - `markdownConverter.Document.HeaderTemplate` and `markdownConverter.Document.FooterTemplate`:
     The html-sourcecode of the header and footer. Variable-substituion is supported here as well. Page-numbers and similar information can be included as described in the [puppeteer docs](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pagepdfoptions).
-  - `markdownConverter.Document.DefaultStyles`:  
+  - `markdownConverter.Document.DefaultStyles`:
     Allows you to disable the default styles. This might be useful if you want to create your own stylesheet from scratch.
-  - `markdownConverter.Assets.StyleSheets`:  
+  - `markdownConverter.Assets.StyleSheets`:
```
[the remark this answers is not in the document]
Comment (L3houx) [reply README.md]: Commit Resolved: **PR README Modifications** ( 9091c20bcc92ddd510cab427289098c1b3a4db51 )

File: src/System/Tasks/ConversionRunner.ts (hunk 2 above)
Comment (L3houx) [reply src/System/Tasks/ConversionRunner.ts]: Commit Resolved: **PR Modifications ConversionRunner TS** ( 7d51275bb9492a485000963129e1df40a3f7b6ca )

File: src/System/Tasks/ConversionRunner.ts (hunk 3 above)
Comment (L3houx) [reply src/System/Tasks/ConversionRunner.ts:523]: Commit Resolved: **PR Modifications ConversionRunner TS** ( 7d51275bb9492a485000963129e1df40a3f7b6ca )

File: package-lock.json (hunk 4 above)
Comment (L3houx) [reply package-lock.json:4]: Commit Resolved: *PR Modifications Package Lock Json* ( d28ede37d4e09345c9f30da013b047c0021f09cc )

File: package.json [hunk 16]
```diff
@@ -70,9 +70,11 @@
     "markdown-it": "^12.2.0",
     "markdown-it-anchor": "^8.3.1",
     "markdown-it-checkbox": "^1.1.0",
+    "markdown-it-container": "^3.0.0",
     "markdown-it-emoji": "^2.0.0",
     "markdown-it-table-of-contents": "^0.5.2",
     "multi-integer-range": "^4.0.9",
+    "path": "^0.12.7",
```
[the remark this answers is not in the document]
Comment (L3houx) [reply package.json:77]: Commit Resolved: *PR Modifications Package Lock Json* ( d28ede37d4e09345c9f30da013b047c0021f09cc )

**Comment by L3houx** [comment]
Thank a lot by the way to take the time to describe them all like this, it's really appreciated.  Once I did all you're recommandations, I update npm and nodejs to latests versions
node -v: `v17.2.0`
npm -v: `8.1.4`

I had this error, Locally:



Same as Build:
image




I was able to fix it using `hashFunction: "xxhash64"` in the webpack config. I will wait your feedback and if needed, I will push back the webpack config

**Comment by manuth** [comment]
@L3houx for some reason I don't know, the Drone CI build didn't start.
Anyways - there are still some linting issues. Could you please have a look at the output of `npm run lint`?

Thanks for taking care 😄

**Comment by L3houx** [comment]
`npm run lint`
image

I will check to fix these warnings

**Comment by L3houx** [comment]
Hi @manuth, I fixed the `npm run lint`

image

I will push my modifications soon.

**Comment by L3houx** [comment]
I have the same error as before, I'm waiting your call to update the webpack config to remove this error.

image

**Comment by manuth** [comment]
Oh I'm sorry - I totally forgot about that.
I'll gladly accept your changes made to the webpack config, then 😄

**Comment by manuth** [comment]
@L3houx running the tests failed.
I assume it's cause of this piece of code:
https://github.com/manuth/MarkdownConverter/blob/810fe455b56d1d4552130f62a529a0096899761a/src/System/Tasks/ConversionRunner.ts#L525

However - I wanted to change how detecting the root works anyways.
If you don't mind I'll push my proposed changes directly to your PR.

**Comment by L3houx** [comment]
It's all good for me, I will wait your commit to use it in the **markdown-it-include** extension configuration

**Comment by manuth** [comment]
@L3houx the CI tests will still fail
This is related to vscode applying IDs to headings by itself.

As this issue is unrelated to this very PR is ready to go.

Let me know if you are ready to have this PR merged or if you have something more to add.

**Comment by L3houx** [comment]
It's all good for me :) You can press the merge button :sunglasses:

**Comment by manuth** [comment]
Alright - did so! Thanks for your contribution, I really appreciate it 😄

**Pull request merged**